text/0048-synchronizer-redesign.md

```diff
@@ -0,0 +1,170 @@
+
+However, we also want to handle cases as well as possible where messages got delayed, or, crucially, where a new node is joining the network and has to catch up and download the full protocol state of the current era. It also has to sometimes wait on the block validator component to check whether an incoming proposal contains a valid block.
+
+The Highway implementation always broadcasts a new vertex if we just created it ourselves. It also always responds to requests for vertices from other nodes.
+
+The synchronizer's job is to decide when to request dependencies, and from which peer. It has to find a compromise between:
+
+* avoiding redundant downloads of the same vertex, and spamming a single peer with too many requests, and
+* downloading dependencies as fast as possible to enable quick finalization of blocks.
+
+The current implementation started off as a naive way of doing this, and has received some isolated improvements, but still makes redundant requests sometimes and is not as efficient as it could be.
+
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The synchronizer keeps collections of pending vertices that cannot be added to the protocol state yet, because:
+
+* their timestamp is from the future,
+* they are missing a dependency, or
+* they are waiting for validation of the block they contain.
+
+Based on these collections, and the peers we received them from, we can deduce which missing dependencies we need to request, and whom we expect to have them: a correct node will only send us a vertex if it already has all dependencies.
+
+To keep the synchronizer's memory usage small and its computations fast, we want to keep its state as small as possible, so we want to prioritize vertices in a way that minimizes the time we expect them to remain pending in the synchronizer: We always want to request those vertices first which are most likely to be added to the protocol state soon. That roughly means: first all evidence against faulty validators, since evidence never has a dependency and can always be added to the protocol state right away; then all endorsements and units, starting with those with an early timestamp, since they tend to have fewer dependencies.
+
+Based on that prioritization we want to send requests so that:
+
+* We request each vertex from only one peer at a time, but with an aggressive timeout.
+* We don't request too many vertices from the same peer at once. The number will be configurable. A low default — even 1 or 2 — should be fine for a running node, but we should determine experimentally whether a higher value speeds up syncing for a joining node.
```

Comment:
I'm not sure about the 1-2 node limit, if it prevents us from queuing all the necessary fetches, we will introduce a round-trip worth of latency to the process each time.

As a first step, we could adjust the weight to get an even distribution, and when a request fails, we redistribute all outstanding ones to other peers?

Another idea is to support batching, requesting more than one vertex at the same time from a single peer and supporting multiple in the response. The peer would reply with both vertices and acknowledges that some are missing, and is allowed to do multiple replies. We would need to keep the message size reasonable though.

Finally is it possible to describe a whole subgraph through one missing vertex maybe? Some like a "get vertex V and all of its dependencies (up to a depth/deliminated by some known hash I am providing as well)"?

All in all, can we lower the number of roundtrips on this by either requesting more in parallel, batching and/or letting the sender figure out what we need?

Reply:
As discussed, sending multiple requests while others are still in flight has roughly the same effect as batching, on the network level. We'll make the limit for parallel requests to the same peer configurable, so we can test with a high value.

> describe a whole subgraph through one missing vertex maybe

I'd like to leave that for later, as a possible improvement: Getting this right and considering all edge cases (equivocations, endorsements, …) will be pretty complicated. Also, if we request too much at once, the probability increases that another peer will send us some of the same information in the meantime.